lib/activerecord_hoarder/batch.rb

```diff
@@ -2,25 +2,44 @@ module ::ActiverecordHoarder
   class Batch
     RECORD_DATE_FIELD = "created_at"
 
-    def self.from_records(record_data)
-      record_data.present? ? new(record_data) : nil
-    end
-
-    def initialize(record_data)
+    def initialize(record_data, delete_transaction: nil)
       @record_data = record_data
       @serializer = ::ActiverecordHoarder::Serializer
+      @delete_transaction = delete_transaction
+    end
+
+    def content_string
+      @serializer.create_archive(@record_data)
     end
 
     def date
-      @date ||= @record_data.first[RECORD_DATE_FIELD].to_date
+      @date ||= extract_date
+    end
+
+    def delete_records!
+      raise(NameError, "batch instantiated without delete transaction") if !@delete_transaction.present?
+      @delete_transaction.call
+    end
+
+    def extract_date
+      return nil if !@record_data.first.present?
+      @record_data.first[RECORD_DATE_FIELD].to_date
     end
 
     def key
       @key ||= ::ActiverecordHoarder::StorageKey.from_date(date, @serializer.extension)
     end
 
-    def content_string
-      @serializer.create_archive(@record_data)
+    def present?
+      @record_data.present?
+    end
+
+    def valid?
+      return false if !present?
+      @record_data.each do |record|
```

Comment:
Wouldn't it be faster to query for a nil vs iterating through thousands of records? If you iterate in ruby, all the SQL record data has to be passed over the wire from database to ruby and serialized into model classes. If you execute a query, you only pass a boolean result set.

Reply:
Makes sense to me, but before venturing into creating a large addition to this changeset, do you think this should rather be a future story? The story would be to move all queries to the batch